generator/views/datagroups.py

```diff
@@ -0,0 +1,162 @@
+    if view.project not in ("moz-fx-data-shared-prod", "mozdata"):
+        logging.debug(
+            f"Unable to get sources for non shared-prod/mozdata view: {view.full_table_id} in generated-sql."
+        )
+        return None
+
+    dataset_view_references = dataset_view_map.get(view.dataset_id)
+    if dataset_view_references is None:
+        logging.debug(f"Unable to find dataset {view.dataset_id} in generated-sql.")
+        return None
+
+    view_references = dataset_view_references.get(view.table_id)
+    if not view_references or len(view_references) > 1:
+        # For views with multiple sources it's unclear which table to check for updates.
+        logging.debug(
+            f"Unable to find a single source for {view.full_table_id} in generated-sql."
+        )
+        return None
+
+    source_table_id = ".".join(view_references[0])
+    try:
+        table = client.get_table(source_table_id)
+        if "TABLE" == table.table_type:
+            return _get_datagroup_from_bigquery_table(table)
+        elif "VIEW" == table.table_type:
+            return _get_datagroup_from_bigquery_view(table, client, dataset_view_map)
+    except NotFound as e:
+        raise ValueError(
+            f"Unable to find {source_table_id} referenced in {view.full_table_id}"
+        ) from e
+
+    return None
+
+
+def _generate_view_datagroup_lkml(
+    view: View,
+    client: bigquery.Client,
+    dataset_view_map: Dict[str, BQViewReferenceMap],
+) -> str:
+    """Generate the Datagroup LookML for a Looker View."""
+    # Only generate datagroup for views that can be linked to a BigQuery table:
+    if view.view_type != TableView.type:
+        return ""
+
+    # Use the release channel table or the first available table (usually the only one):
+    view_table = next(
+        (table for table in view.tables if table.get("channel") == "release"),
+        view.tables[0],
+    )["table"]
+
+    try:
+        bq_table = client.get_table(view_table)
+    except NotFound as e:
+        raise ValueError(
+            f"{view_table} not found when generating datagroup but in namespaces yaml."
+        ) from e
+
+    if "TABLE" == bq_table.table_type:
+        datagroup: Optional[Datagroup] = _get_datagroup_from_bigquery_table(bq_table)
+        return str(datagroup)
+    elif "VIEW" == bq_table.table_type:
+        datagroup = _get_datagroup_from_bigquery_view(
+            bq_table, client, dataset_view_map
+        )
+        if datagroup is not None:
+            return str(datagroup)
+
+    return ""
+
+
+def generate_datagroups(
+    views: List[View], target_dir: Path, namespace: str, client: bigquery.Client
+) -> None:
+    """Generate and write a datagroups.lkml file to the namespace folder."""
+    datagroups_lkml_path = target_dir / namespace / "datagroups.lkml"
+
+    # To map views to their underlying tables:
+    dataset_view_map = lookml_utils.get_bigquery_view_reference_map(
+        DEFAULT_GENERATED_SQL_URI
```

Comment:
Elsewhere this is a CLI parameter but I've just used the default here for simplicity. If there's a reason not do this I can pass this in as a param in the upstream `lookml` function